src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -102,29 +102,54 @@ static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
             static bool NeedsEscaping(char c) => c is '[' or ']' or '&' or '*' or ',' or '+' or EscapeCharacter;
         }
 
-        internal static ReadOnlySpan<char> GetName(ReadOnlySpan<char> fullName)
+        internal static ReadOnlySpan<char> GetNamespace(ReadOnlySpan<char> fullName)
         {
-            // The two-value form of MemoryExtensions.LastIndexOfAny does not suffer
-            // from the behavior mentioned in the comment at the top of GetFullTypeNameLength.
-            // It always takes O(m * i) worst-case time and is safe to use here.
+            // Detect if a nested type name has a namespace (e.g. "a.b+c") and throw if it does.
+            while (LastIndexOfUnescaped(fullName, '+') is int nestingBoundaryPos && nestingBoundaryPos > 0)
+            {
+                if (LastIndexOfUnescaped(fullName.Slice(nestingBoundaryPos + 1), '.') > 0)
+                {
+                    ThrowInvalidOperation_NestedTypeNamespace();
+                }
+                fullName = fullName.Slice(0, nestingBoundaryPos);
+            }
+            int offset = LastIndexOfUnescaped(fullName, '.');
+            return offset < 0 ? [] : fullName.Slice(0, offset);
+        }
 
-            int offset = fullName.LastIndexOfAny('.', '+');
+        internal static ReadOnlySpan<char> GetName(ReadOnlySpan<char> fullName)
+        {
+            // If the type is nested, return the whole name after the plus sign.
+            int offset = LastIndexOfUnescaped(fullName, '+');
 
-            if (offset > 0 && fullName[offset - 1] == EscapeCharacter) // this should be very rare (IL Emit & pure IL)
+            if (offset < 0)
             {
-                offset = GetUnescapedOffset(fullName, startIndex: offset);
+                // Look for dots only if the type is not nested.
+                offset = LastIndexOfUnescaped(fullName, '.');
```

Comment:
dots are never escaped. Escaped dot should have been rejected by the parser earlier. I think this can use regular last index of.

Reply:
Done. This removes a couple test cases.